File: HW/HW1/HW1.py
# brute force (two for loops)
def isPalindrome1(self, s): 
    s = s.lower()
    editted = ""

    for c in s:
        if c.isalnum():
            editted += c
        
    for i in range(len(editted) / 2):
        if not editted[i] == editted[len(editted) - i - 1]: 
            return False
    
    return True


# two pointer
def isPalindrome2(s: str) -> bool:
    p1 = 0
    p2 = len(s) - 1
    s = s.lower()

    while (p1 < p2):
        if not s[p1].isalnum():
            p1 += 1
        elif not s[p2].isalnum():
            p2 -= 1
        elif s[p1] == s[p2]:
            p1 += 1
            p2 -= 1
        else: 
            return False
        
    return True


words = ["A man, a plan, a canal: Panama", "racecar", "race a car", " "]

File: HW/HW1/HW1.java
package HW.HW1;

public class HW1 {
    //brute force (two for loops)
    public boolean isPalindrome1(String s) {
        s = s.toLowerCase();
        String edited = ""; 
        for (int i = 0; i < s.length(); i++) { 
            if (Character.isLetterOrDigit(s.charAt(i))) { 
                edited += s.charAt(i);
            }
        }
        if (edited.length() <= 1) { 
            return true;
        }
        

        for (int i = 0; i < edited.length() / 2; i++) { 
            if (edited.charAt(i) != edited.charAt(edited.length() - i - 1)) { 
                return false;
            }
        }
        return true;
    }

    //two pointer
    public boolean isPalindrome2(String s) {
        s = s.toLowerCase();
        int p1 = 0; 
        int p2 = s.length() - 1;

        while (p1 <= p2) { 
            char c1 = s.charAt(p1);
            char c2 = s.charAt(p2);

            if (!Character.isLetterOrDigit(c1)) { 
                p1++;
            } else if (!Character.isLetterOrDigit(c2)) { 
                p2--;
            } else { 
                if (c1 != c2) { 
                    return false; 
                }
                p1++;
                p2--;
            }
        }
        return true;
    }
}    

